readme.txt
=== Defender Security, Monitoring, and Hack Protection ===
Plugin Name: Defender Security, Monitoring, and Hack Protection
Version: 1.6.2
Author: WPMU DEV
Author URI: http://premium.wpmudev.org/
Contributors: WPMUDEV
Tags: Security, Security Tweaks, Hardening, IP lockout, Monitoring, Blacklist, Site Protection, Hacked, Security Scan
Requires at least: 3.5
Tested up to: 4.8
Stable tag: 1.6.2
License: GPL v2 - http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

Protect WordPress from hackers with security tweaks, code scans, IP lockouts, and monitoring.



== Description ==

<strong>Defender is layered security for WordPress made easy.</strong> And by easy, we mean <strong>amazingly easy!</strong> No longer do you have to go through hideously complex settings and get a virtual PhD in security. Defender adds all the hardening and security tweaks you need, in minutes. :)

= Security Tweaks =

Defender starts with a list of one-click hardening techniques that will instantly add layers of protection to your site.

= Block hackers at every level: =

* Disable trackbacks and pingbacks - safety first
* Core and server update recommendations - stay on top of your systems
* Change default database prefix - they won’t find this
* Disable file editor - if they get in, they won’t get far
* Hide error reporting - don’t reveal your issues
* Update security keys - ultimate security reset
* Prevent information disclosure - why tell them what you have
* Prevent PHP execution - because it’s daaaangerous

= File Scans =

Run free scans that check WordPress for suspicious code. The Defender scan tool compares your WordPress install with the directory, reports changes and lets you restore the original file with a click.
 
= IP Blacklist =
Keep your site safe with Defender’s simple IP manager. Manually block specific IPs, import a list of banned IPs and set automated timed and permanent lockouts. Defender makes it easy to quickly block and unblock specific locations.

★★★★★ <br>
“I found other pro security plugins a bit too fiddly for my taste...I’m delighted with Defender” - <a href="https://profiles.wordpress.org/keithadv">KeithADV</a>

★★★★★ <br>
“Defender's interface is very intuitive with warnings that are very helpful” - <a href="https://premium.wpmudev.org/profile/djohns">djohns</a>

= Log-in Protection =

Brute force attacks are no match for Defender. Limit login attempts to stop users trying to guess passwords. Permanently ban IPs or trigger a timed lockout after a set number of failed login attempts.

= 404 Limiter =
Defender detects when bots are being used to scan your site for vulnerabilities and shuts them down. The 404 limiter lets you stop the scan by detecting when a user keeps visiting pages that do not exist.

= Notifications and Reports =
Defender runs surveillance and sends notifications with information that matters.

= Features Available in Defender Include: =

* One-click site hardening and security tweaking
* WordPress core file scanning and repair
* IP Blacklist manager and logging
* Unlimited file scans
* Timed Lockout brute force attack shield for login protection 
* 404 limiter for blocking vulnerability scans
* IP lockout notifications and reports

Defender can take care of all your security needs, for free!

However, if you'd like extra scanning, audits and monitoring, you can always take the next step with Defender Pro.



== Frequently Asked Questions ==

= Why should I choose Defender over other security plugins? =

Defender is built to add all the best hardening and security tweaks used by the pros without having to become a security expert. This means you get all the most effective and proven protection methods other services provide with fewer settings, on-click hardening and faster setup.

= Is Defender the only step I need to take in securing my WordPress site? =
Hackers and bot attacks are not the only threat to your site.  No matter what security plugin or service you use, always be prepared with a secure backup stored in a safe location away from your live site. Security does not protect from hosting outages, server errors and accidentally lost or damaged data. We recomend <a href="https://premium.wpmudev.org/project/snapshot/">Snapshot</a>. Defender with scheduled managed backups is the best way to keep your site safe.



== Screenshots ==

1. Layered security tweaks let your harden your site with a few clicks.
2. Compares your WordPress install with the directory and restore original files with a click.
3. Simple configuration and security manager.
4. IP blacklisting, 404 limiter and Timed Lockout attack shield



== Installation ==

1. Upload the `wp-defender` plugin to your `/wp-content/plugins/` directory.
1. Activate the plugin through the 'Plugins' menu in WordPress.
1. Configure and manage using the `defender` menu item in the WordPress dashboard.
1. Done!



== Changelog ==

= 1.6.2 =
* New: CSV export for Audit Logging.
* Improvement: Email reports now have unsubscribe link, and link to Reports where email reports can be turned off.
* Fix: Typo in Audit email.
* Other minor enhancements/fixes


= 1.6.1 =
* Improvement: Improved IP Lockout performance.
* Fix: "Update old security keys" doesn't move to resolved list after processed
* Fix: When emptying IP Lockout logs cause timeout error.
* Fix: Typos in some places
* Other minor enhancements/fixes

= 1.6.0.1 =
* This is a quick hotfix release to tame a few notifications that showed up when they weren't supposed to. Joy.

= 1.6 =
* New: First edition for WordPress directory

= 1.5 =
* New: You can now add exceptions for specific PHP files in the PHP Execution Security Tweak.
* Improvement: Filtering all log types now uses URLs instead of ajax only, meaning you can link to a filtered log easily.
* Improvement: Various user experience updates across the plugin interface to make using Defender even easier.
* Fix: Lockout Logs now display from newest to oldest.
* Fix: Lockout Logs pagination now works correctly.
* Fix: Inconsistencies in the IP Lockouts stats across the plugin.
* Fix: Sending Audit Logging reports to multiple recipients would address all recipients as the first user's name.
* Fix: Grammar and typos in some modals and error messages.
* Fix: If Defender finds a vulnerability in WordPress's core, the text would indicate running an update would fix the issue though no update was actually available yet.

= 1.4.2 =
* Improvement: The plugin interface will now stretch to utilize extra screen space on larger screens.
* Fix: Audit Logging was getting its days mixed up in the summary area. You’ll now see the correct day of the week.
* Fix: We squashed a bug that was causing files scans to sometimes report false positive files after WordPress core upgrades.
* Fix: A conflict with Jetpack was causing scans to stall, which we have now fixed up.
* Fix: In some cases File Scanning reports wouldn't actually stop sending if you disabled them. It now obeys commands.
* Fix: Google's bot was being blocked by IP Lockouts but now it's free to crawl and index as it pleases.
* Fix: We removed redundant “cancel” buttons on settings pages. You probably won’t even notice!
* Fix: We’ve added live stats so now there’s no need to wait around in anticipation while running files scan actions.
* Fix: Stats weren’t displaying the right numbers after actioning security tweaks, but it’s all good now.
* Fix: Pagination on the Audit Logging logs page now works like you would expect it to.
* Fix: Files detected in File Scanning now have metrics with their file sizes.
* Fix: We’ve fixed styling issues with toggles.
* Fix: We removed the” Resolve bulk update” option from File Scanning. It wasn’t really a valid action.
* Fix: Incomplete icons in the Dashboard reports area have been updated.
* Fix: We’ve removed redirection from the dashboard to the File Scanning page are after preforming a file scan so now you shouldn’t feel lost.
* Fix: Lots of other small stuff, like minor cosmetic and grammar fixes.

= 1.4.1 =
* Fixed: Compatibility issue with Getting Started Wizard
* Fixed: Scanning was sometimes slow or getting stuck

= 1.4 =
* New: Meet the brand new Defender! This release focuses on making security for WordPress a better place. We’ve given the UI a refresh and updated the UX, so configuring your security settings is a walk in the park.
* Fixed: A ton of bug fixes & improvements. Yep, vague description! But why bore you with the small stuff when you could be spending time bolstering your site’s security?

= 1.3 =
* Added: Endpoint API so HUB can work with Defender natively through WPMU DEV Dashboard plugin
* Other minor enhancements/fixes

= 1.2 =
* Added: New Hardening Rule (PHP version)
* Improvement: Audit Logging now allows date range selection.
* Improvement: IP Lockouts now allow IP ranges in whitelist/blacklist.
* Improvement: IP Lockouts now can import/export whitelist/backlist.
* Fixed: IP Lockouts email notification text on permanent IP ban.

= 1.1.4.1 =
* Fixed: Fatal error when PHP extension sockets is not enabled

= 1.1.4 =
* Improvement: Audit logging now detects file changes in WordPress core.
* Fixed: Updating via WordPress core now syncs better with the Hub.
* Fixed: Some compatibility fixes for PHP 5.2.

= 1.1.3 =
* Improvement: Audit Logging now ajax based.
* Fixed: minor bug fixes & some UI/UX improvements

= 1.1.2 =
* Improvement: Switched the User dropdown in Audit Logging to load results via AJAX to increase initial load performance.
* Improvement: Scan results now pre-load information so that you can action fixes faster.
* Fixed: Removed cronjob events from being tracked in Audit Logging.
* Fixed: The Audit Logging filter box now stays visible if no results are returned.
* Fixed: Other small bug fixes and improvements.

= 1.1.1 =
* Added: A warning indicator in WP Admin sidebar to let you know how many security issues are outstanding.
* Added: The ability to choose to only receive email reports when there are issues with your website.
* Fixed: Minor bug fixes & improvements

= 1.1 =
* New feature: Audit logging
* New plugin icon
* Vulnerability plugins/theme scan result can be ignored
* Some other minor enhancements/fixes

= 1.0.8 =
* Improve Core Integrity Scan.
* Improve caching method

= 1.0.7 =
* Improved: Scan schedule.
* Fix: issue with W3 Total Cache Object Cache

= 1.0.6 =
* Fix: Defender data doesn't sync with HUB correctly
* Fix: Email report doesn't send properly
* Some other minor enhancements/fixes

= 1.0.5 =
* Added: Option to choose reminder period for Hardener rule "Update old security keys"
* Improved: Compatibility with Windows server
* Improved: Optimized resource usage when scanning
* Fix: issue with memcached
* Other minor enhancements/fixes

= 1.0.4 =
* Improve scan engine, reduce false positives
* Improve uninstallation method
* Add the ability to ignore hardener rules.
* Improve the performance impact on the site.
* Fix scans sticking at 100% in some cases
* Fix compatibility issues with IIS
* Some other minor enhancements/fixes

= 1.0.3 =
* Optimize scanning
* Preventing performance issue with some hosts
* Fix false blacklist detection in local environment
* Some other minor enhancements/fixes

= 1.0.2 =
* Applied ajax inline updates for plugins/themes
* One click Prevent PHP execution
* One click Prevent Information Disclosure
* Add detail page for core integrity issue, and automate resolution
* Fix scan stability with limited memory
* Some other minor enhancements/fixes

= 1.0.1 =
* Scanning can auto detect if user is active on scanning page to work based on ajax, or leave to enable background scan
* Improve condition checking for Prevent Information Disclosure module
* Improve condition checking for Prevent PHP execution module

= 1.0 = 
* First release


== About Us ==
WPMU DEV is a premium supplier of quality WordPress plugins and themes. For premium support with any WordPress related issues you can join us here:
<a href="https://premium.wpmudev.org/?utm_source=wordpress.org&utm_medium=readme">https://premium.wpmudev.org/</a>

Don't forget to stay up to date on everything WordPress from the Internet's number one resource:
<a href="https://premium.wpmudev.org/blog/?utm_source=wordpress.org&utm_medium=readme">WPMU DEV Blog</a>

Hey, one more thing... we hope you <a href="http://profiles.wordpress.org/WPMUDEV/">enjoy our free offerings</a> as much as we've loved making them for you!

changelog.txt
Plugin Name: WP Defender
Author: Hoang Ngo, Aaron Edwards
Tested up to: 4.7.4

Change Log:

1.6.2 - 2017-05-07
----------------------------------------------------------------------
- Improvement: Email reports now have unsubscribe link, and link to Reports where email reports can be turned off.
- Other minor enhancements/fixes

1.6.1 - 2017-21-06
----------------------------------------------------------------------
- Improvement: Improved IP Lockout performance.
- Fix: Audit logging detects wrong WordPress version when upgrade
- Fix: "Update old security keys" doesn't move to resolved list after processed
- Fix: When emptying IP Lockout logs cause timeout error.
- Fix: Typos in some places
- Other minor enhancements/fixes

1.6 - 2017-05-06
----------------------------------------------------------------------
- Improvement: Allow users to select and apply rules to other server type in Prevent PHP Execution and Prevent Information Disclosure.
- Fix: Sometimes HUB status doesn't sync with WordPress site.
- Other minor enhancements/fixes

1.5 - 2017-17-05
----------------------------------------------------------------------
- New: You can now add exceptions for specific PHP files in the PHP Execution Security Tweak.
- Improvement: Filtering all log types now uses URLs instead of ajax only, meaning you can link to a filtered log easily.
- Improvement: Various user experience updates across the plugin interface to make using Defender even easier.
- Fix: Lockout Logs now display from newest to oldest.
- Fix: Lockout Logs pagination now works correctly.
- Fix: Inconsistencies in the IP Lockouts stats across the plugin.
- Fix: Sending Audit Logging reports to multiple recipients would address all recipients as the first user's name.
- Fix: Grammar and typos in some modals and error messages.
- Fix: If Defender finds a vulnerability in WordPress's core, the text would indicate running an update would fix the issue though no update was actually available yet.

1.4.2 - 2017-03-05
----------------------------------------------------------------------
- Improvement: The plugin interface will now stretch to utilize extra screen space on larger screens.
- Fix: Audit Logging was getting its days mixed up in the summary area. You’ll now see the correct day of the week.
- Fix: We squashed a bug that was causing files scans to sometimes report false positive files after WordPress core upgrades.
- Fix: A conflict with Jetpack was causing scans to stall, which we have now fixed up.
- Fix: In some cases File Scanning reports wouldn't actually stop sending if you disabled them. It now obeys commands.
- Fix: Google's bot was being blocked by IP Lockouts but now it's free to crawl and index as it pleases.
- Fix: We removed redundant “cancel” buttons on settings pages. You probably won’t even notice!
- Fix: We’ve added live stats so now there’s no need to wait around in anticipation while running files scan actions.
- Fix: Stats weren’t displaying the right numbers after actioning security tweaks, but it’s all good now.
- Fix: Pagination on the Audit Logging logs page now works like you would expect it to.
- Fix: Files detected in File Scanning now have metrics with their file sizes.
- Fix: We’ve fixed styling issues with toggles.
- Fix: We removed the” Resolve bulk update” option from File Scanning. It wasn’t really a valid action.
- Fix: Incomplete icons in the Dashboard reports area have been updated.
- Fix: We’ve removed redirection from the dashboard to the File Scanning page are after preforming a file scan so now you shouldn’t feel lost.
- Fix: Lots of other small stuff, like minor cosmetic and grammar fixes.

1.4.1 - 2017-21-04
----------------------------------------------------------------------
Fixed: Compatibility issue with Getting Started Wizard
Fixed: Scanning was sometimes slow or getting stuck

1.4 - 2017-18-04
----------------------------------------------------------------------
- New: Meet the brand new Defender! This release focuses on making security for WordPress a better place. We’ve given the UI a refresh and updated the UX, so configuring your security settings is a walk in the park.
- Fixed: A ton of bug fixes & improvements. Yep, vague description! But why bore you with the small stuff when you could be spending time bolstering your site’s security?

1.3 - 2017-13-03
----------------------------------------------------------------------
- Added: Endpoint API so HUB can work with Defender natively through WPMU DEV Dashboard plugin
- Other minor enhancements/fixes

1.2 - 2017-27-02
----------------------------------------------------------------------
- Added: New Hardening Rule (PHP version)
- Improvement: Audit Logging now allows date range selection.
- Improvement: IP Lockouts now allow IP ranges in whitelist/blacklist.
- Improvement: IP Lockouts now can import/export whitelist/backlist.
- Fixed: IP Lockouts email notification text on permanent IP ban.

1.1.4.1 - 2016-31-10
----------------------------------------------------------------------
- Fixed: Fatal error when PHP extension sockets is not enabled

1.1.4 - 2016-31-10
----------------------------------------------------------------------
- Improvement: Audit logging now detects file changes in WordPress core.
- Fixed: Updating via WordPress core now syncs better with the Hub.
- Fixed: Some compatibility fixes for PHP 5.2.

1.1.3 - 2016-20-09
----------------------------------------------------------------------
- Improvement: Audit Logging now ajax based.
- Fixed: minor bug fixes & some UI/UX improvements

1.1.2 - 2016-24-08
----------------------------------------------------------------------
- Improvement: Switched the User dropdown in Audit Logging to load results via AJAX to increase initial load performance.
- Improvement: Scan results now pre-load information so that you can action fixes faster.
- Fixed: Removed cronjob events from being tracked in Audit Logging.

app/behavior/utils.php
<?php
/**
 * Author: Hoang Ngo
 */

namespace WP_Defender\Behavior;

use Hammer\Base\Behavior;
use Hammer\Helper\Log_Helper;
use Hammer\Helper\WP_Helper;
use WP_Defender\Module\Hardener\Model\Settings;
use WP_Defender\Module\Scan\Component\Scan_Api;
use WP_Defender\Module\Scan\Model\Result_Item;

class Utils extends Behavior {
	/**
	 * @return bool
	 */
	public function isActivatedSingle() {
		if ( WP_Helper::is_network_activate( wp_defender()->plugin_slug ) ) {
			return false;
		}

		return true;
	}

	/**
	 * @param $endPoint
	 * @param array $bodyArgs
	 * @param array $requestArgs
	 * @param bool $returnRaw
	 *
	 * @return array|mixed|object|\WP_Error
	 */
	public function devCall( $endPoint, $bodyArgs = array(), $requestArgs = array(), $returnRaw = false ) {
		$api_key = $this->getAPIKey();
		if ( $api_key !== false ) {
			$domain                      = network_site_url();
			$post_vars['body']           = $bodyArgs;
			$post_vars['body']['domain'] = $domain;
			$post_vars['timeout']        = 30;
			$post_vars['httpversion']    = '1.1';

			$headers = isset( $post_vars['headers'] ) ? $post_vars['headers'] : array();

			$post_vars['headers']               = array_merge( $headers, array(
				'Authorization' => 'Basic ' . $api_key
			) );
			$post_vars['headers']['user-agent'] = 'http://asdc.com';

			$post_vars = array_merge( $post_vars, $requestArgs );

			$response = wp_remote_request( $endPoint,
				apply_filters( 'wd_wpmudev_call_request_args',
					$post_vars ) );

			if ( is_wp_error( $response ) ) {
				return $response;
			}

			if ( $returnRaw == true ) {
				return $response;
			}

			if (
				'OK' !== wp_remote_retrieve_response_message( $response )
				OR 200 !== wp_remote_retrieve_response_code( $response )
			) {
				return new \WP_Error( wp_remote_retrieve_response_code( $response ), wp_remote_retrieve_response_message( $response ) );
			} else {
				$data = wp_remote_retrieve_body( $response );

				return json_decode( $data, true );
			}
		} else {
			return new \WP_Error( 'dashboard_required',
				sprintf( esc_html__( "WPMU DEV Dashboard will be required for this action. Please visit <a href=\"%s\">here</a> and install the WPMU DEV Dashboard", wp_defender()->domain )
					, 'https://premium.wpmudev.org/project/wpmu-dev-dashboard/' ) );
		}
	}

	/**
	 * @return string
	 */
	public function getPHPVersion() {
		return phpversion();
	}

	/**
	 * @return string
	 */
	public function getWPVersion() {
		global $wp_version;

		return $wp_version;
	}

	public function getAPIKey() {
		if ( ( $version = $this->isDashboardInstalled() ) !== false ) {
			if ( version_compare( $version, '4.0.0' ) >= 0 ) {
				//this is version 4+
				//instanize once
				\WPMUDEV_Dashboard::instance();
				$api_key = \WPMUDEV_Dashboard::$api->get_key();

				return $api_key;
			} else {
				global $wpmudev_un;
				$api_key = $wpmudev_un->get_apikey();

				return $api_key;
			}
		}

		return false;
	}

	/**
	 * Check if WPMUDEV Dashboard installed, return version, else return false
	 * @return bool|string
	 */
	public function isDashboardInstalled() {
		//check if this is new
		if ( class_exists( 'WPMUDEV_Dashboard' ) ) {
			return \WPMUDEV_Dashboard::$version;
		}

		return false;
	}

	/**
	 * Check if current user qualify for current operation
	 * @return bool
	 */
	public function checkPermission() {
		if ( ! is_user_logged_in() ) {
			return false;
		}
		$cap = is_multisite() ? 'manage_network_options' : 'manage_options';

		return current_user_can( $cap );
	}

	/**
	 * @param $timestring string
	 * @param bool $i18n
	 *
	 * @return false|string
	 */
	public function formatDateTime( $timestring, $i18n = true ) {
		if ( strlen( $timestring ) == 0 ) {
			return null;
		}
		if ( ! is_int( $timestring ) ) {
			$timestring = strtotime( $timestring );
		}
		$format = get_option( 'date_format' ) . ' ' . get_option( 'time_format' );
		if ( $i18n == false ) {
			return date( $format, $timestring );
		} else {
			$time = get_date_from_gmt( date( 'Y-m-d H:i:s', $timestring ), 'Y-m-d H:i:s' );

			return date_i18n( $format, strtotime( $time ) );
		}
	}

	/**
	 * @param null $user_id
	 *
	 * @return string
	 */
	public function getDisplayName( $user_id = null ) {
		if ( ! is_user_logged_in() && is_null( $user_id ) ) {
			return esc_html__( "Guest", wp_defender()->domain );
		}

		if ( is_null( $user_id ) ) {
			$user_id = get_current_user_id();
		}

		$userdata = get_userdata( $user_id );
		if ( ! is_object( $userdata ) ) {
			return __( "Guest", wp_defender()->domain );
		}

		$fullname = trim( $userdata->first_name . ' ' . $userdata->last_name );
		if ( empty( $fullname ) ) {
			$fullname = $userdata->display_name;
		}

		return $fullname;
	}

	/**
	 * @return array
	 */
	public function allowedHtml() {
		return array(
			'p'      => array(),
			'i'      => array(
				'class' => 'wd-text-warning wdv-icon wdv-icon-fw wdv-icon-exclamation-sign'
			),
			'strong' => array(),
			'span'   => array(
				'class' => array(
					'wd-suspicious-strong',
					'wd-suspicious-light',
					'wd-suspicious-medium'
				)
			),
			'img'    => array(
				'class' => 'text-warning',
				'src'   => wp_defender()->getPluginUrl() . 'assets/img/robot.png'
			)
		);
	}

	/**
	 * @param $get_avatar
	 *
	 * @return mixed
	 */
	public function getAvatarUrl( $get_avatar ) {
		preg_match( "/src='(.*?)'/i", $get_avatar, $matches );

		return $matches[1];
	}

	/**
	 * Make filesize friendy with human, src from http://jeffreysambells.com/2012/10/25/human-readable-filesize-php
	 *
	 * @param $bytes
	 *
	 * @return string
	 *
	 */
	public function makeReadable( $bytes ) {
		if ( $bytes >= 1073741824 ) {
			$bytes = number_format( $bytes / 1073741824, 2 ) . ' GB';
		} elseif ( $bytes >= 1048576 ) {
			$bytes = number_format( $bytes / 1048576, 2 ) . ' MB';
		} elseif ( $bytes >= 1024 ) {
			$bytes = number_format( $bytes / 1024, 2 ) . ' KB';
		} elseif ( $bytes > 1 ) {
			$bytes = $bytes . ' bytes';
		} elseif ( $bytes == 1 ) {
			$bytes = $bytes . ' byte';
		} else {
			$bytes = '0 bytes';
		}

		return $bytes;
	}

	/**
	 * @param $timestamp
	 *
	 * @return false|int
	 */
	public function localToUtc( $timestring ) {
		$tz = get_option( 'timezone_string' );
		if ( ! $tz ) {
			$gmt_offset = get_option( 'gmt_offset' );
			if ( $gmt_offset == 0 ) {
				return strtotime( $timestring );
			}
			$tz = $this->getTimezoneString( $gmt_offset );
		}
		if ( ! $tz ) {
			$tz = 'UTC';
		}
		$timezone = new \DateTimeZone( $tz );
		$time     = new \DateTime( $timestring, $timezone );

		return $time->getTimestamp();
	}

	/**
	 * @param $time string - format H:i
	 * @param $hook string - hook
	 *
	 * @return false|int
	 */
	public function reportCronTimestamp( $time, $hook ) {
		wp_clear_scheduled_hook( $hook );
		$timeString = date( 'Y-m-d', current_time( 'timestamp' ) ) . ' ' . $time . ':00';
		$timestamp  = $this->localToUtc( $timeString );
		if ( $timestamp > time() ) {
			return $timestamp;
		} else {
			//time is passed, tomorrow
			$timeString = date( 'Y-m-d', strtotime( 'tomorrow', current_time( 'timestamp' ) ) ) . ' ' . $time . ':00';

			return $this->localToUtc( $timeString );
		}
	}

	/**
	 * @param $interval
	 * @param $day
	 * @param $lastReportTime
	 *
	 * @return bool
	 */
	public function isReportTime( $interval, $day, $lastReportTime = false ) {
		if ( $interval == 1 ) {
			//this is daily, always send when interval come
			return true;
		}
		$current_day = strtolower( strftime( '%A', current_time( 'timestamp' ) ) );
		if ( $interval == 7 && $current_day == strtolower( $day ) ) {
			//check the day
			return true;
		} elseif ( $interval == 30
		           && $lastReportTime
		           && strtotime( '+30 days', $lastReportTime ) < time()
		           && $current_day == strtolower( $day )
		) {
			return true;
		}

		return false;
	}

	/**
	 * @param $timezone
	 *
	 * @return false|string
	 */
	public function getTimezoneString( $timezone ) {
		$timezone = explode( '.', $timezone );
		if ( isset( $timezone[1] ) ) {
			$timezone[1] = 30;
		} else {
			$timezone[1] = '00';
		}
		$offset = implode( ':', $timezone );
		list( $hours, $minutes ) = explode( ':', $offset );
		$seconds = $hours * 60 * 60 + $minutes * 60;
		$tz      = timezone_name_from_abbr( '', $seconds, 1 );
		if ( $tz === false ) {
			$tz = timezone_name_from_abbr( '', $seconds, 0 );
		}

		return $tz;
	}

	/**
	 * Get days of week
	 * @return mixed|void
	 * @access public
	 * @since 1.0
	 */
	public function getDaysOfWeek() {
		$timestamp = strtotime( 'next Sunday' );
		$days      = array();
		for ( $i = 0; $i < 7; $i ++ ) {
			$days[]    = strftime( '%A', $timestamp );
			$timestamp = strtotime( '+1 day', $timestamp );
		}

		return apply_filters( 'wd_scan_get_days_of_week', $days );
	}

	/**
	 * A shorhand function to get user IP
	 * @return mixed|string
	 */
	public function getUserIp() {
		$client      = isset( $_SERVER['HTTP_CLIENT_IP'] ) ? $_SERVER['HTTP_CLIENT_IP'] : null;
		$forward     = isset( $_SERVER['HTTP_X_FORWARDED_FOR'] ) ? $_SERVER['HTTP_X_FORWARDED_FOR'] : null;
		$remote      = isset( $_SERVER['REMOTE_ADDR'] ) ? $_SERVER['REMOTE_ADDR'] : null;
		$client_real = isset( $_SERVER['HTTP_X_REAL_IP'] ) ? $_SERVER['HTTP_X_REAL_IP'] : null;
		$ret         = $remote;
		if ( filter_var( $client, FILTER_VALIDATE_IP ) ) {
			$ret = $client;
		} elseif ( filter_var( $client_real, FILTER_VALIDATE_IP ) ) {
			$ret = $client_real;
		} elseif ( ! empty( $forward ) ) {
			$forward = explode( ',', $forward );
			$ip      = array_shift( $forward );
			$ip      = trim( $ip );
			if ( filter_var( $ip, FILTER_VALIDATE_IP ) ) {
				$ret = $ip;
			}
		}

		return apply_filters( 'defender_user_ip', $ret );
	}

	/**
	 * Return times frame for selectbox
	 * @access public
	 * @since 1.0
	 */
	public function getTimes() {
		$data = array();
		for ( $i = 0; $i < 24; $i ++ ) {
			foreach ( apply_filters( 'wd_scan_get_times_interval', array( '00', '30' ) ) as $min ) {
				$time          = $i . ':' . $min;
				$data[ $time ] = apply_filters( 'wd_scan_get_times_hour_min', $time );
			}
		}

		return apply_filters( 'wd_scan_get_times', $data );
	}

	/**
	 * a quick helper for static class
	 * @return Utils
	 */
	public static function instance() {
		return new Utils();
	}

	public function determineServer() {
		$url         = home_url();
		$server_type = get_site_transient( 'wd_util_server' );
		if ( ! is_array( $server_type ) ) {
			$server_type = array();
		}

		if ( isset( $server_type[ $url ] ) && ! empty( $server_type[ $url ] ) ) {
			return strtolower( $server_type[ $url ] );
		}

		//url should be end with php
		global $is_apache, $is_nginx, $is_IIS, $is_iis7;

		$server = null;

		if ( $is_nginx ) {
			$server = 'nginx';
		} elseif ( $is_apache ) {
			//case the url is detecting php file
			if ( pathinfo( $url, PATHINFO_EXTENSION ) == 'php' ) {
				$server = 'apache';
			} else {
				//so the server software is apache, let see what the header return
				$request = wp_remote_head( $url, array( 'user-agent' => $_SERVER['HTTP_USER_AGENT'] ) );
				$server  = wp_remote_retrieve_header( $request, 'server' );
				$server  = explode( '/', $server );
				if ( strtolower( $server[0] ) == 'nginx' ) {
					//proxy case
					$server = 'nginx';
				} else {
					$server = 'apache';
				}
			}
		} elseif ( $is_iis7 || $is_IIS ) {
			$server = 'iis';
		}

		if ( is_null( $server ) ) {
			//if fall in here, means there is st unknowed.
			$request = wp_remote_head( $url, array( 'user-agent' => $_SERVER['HTTP_USER_AGENT'] ) );
			$server  = wp_remote_retrieve_header( $request, 'server' );
			$server  = explode( '/', $server );
			$server  = $server[0];
		}

		$server_type[ $url ] = $server;
		//cache for an hour
		set_site_transient( 'wd_util_server', $server_type, 3600 );

		return $server;
	}

	/**
	 * @return string
	 */
	public function getDefUploadDir() {
		$uploadDir = WP_Helper::getUploadDir();
		$defDir    = $uploadDir . DIRECTORY_SEPARATOR . 'wp-defender';
		if ( ! is_dir( $defDir ) ) {
			wp_mkdir_p( $defDir );
		}

		if ( ! is_file( $defDir . DIRECTORY_SEPARATOR . 'index.php' ) ) {
			//create a blank index file
			file_put_contents( $defDir . DIRECTORY_SEPARATOR . 'index.php', '' );
		}

		return $defDir;
	}

	/**
	 * @return array|void
	 */
	public function submitStatsToDev() {
		if ( ! $this->getAPIKey() ) {
			return;
		}

		$issues = array();
		$rules  = Settings::instance()->getIssues();
		foreach ( $rules as $rule ) {
			$issues[] = array(
				'label' => $rule->getTitle(),
				'url'   => network_admin_url( 'admin.php?page=wdf-hardener' ) . '#' . $rule::$slug
			);
		}

		$model = Scan_Api::getLastScan();
		$count = 0;
		if ( is_object( $model ) ) {
			$timestamp = strtotime( $model->dateFinished );

			$res = array(
				'core_integrity'   => $model->getCount( 'core' ),
				'vulnerability_db' => $model->getCount( 'vuln' ),
				'file_suspicious'  => $model->getCount( 'content' )
			);

			$count = $model->countAll( Result_Item::STATUS_ISSUE );
		} else {
			$timestamp = '';
			$res       = array(
				'core_integrity'   => 0,
				'vulnerability_db' => 0,
				'file_suspicious'  => 0
			);
		}
		$labels       = array(
			'core_integrity'   => esc_html__( "WordPress Core Integrity", wp_defender()->domain ),
			'vulnerability_db' => esc_html__( "Plugins & Themes vulnerability", wp_defender()->domain ),
			'file_suspicious'  => esc_html__( "Suspicious Code", wp_defender()->domain )
		);

		$data = array(
			'domain'       => network_home_url(),
			'timestamp'    => $timestamp,
			'warnings'     => $count,
			'cautions'     => count( $issues ),
			'data_version' => '20160220',
			'scan_data'    => json_encode( array(
				'scan_result'        => $res,
				'hardener_result'    => $issues,
				'scan_schedule'      => array(
					'is_activated' => \WP_Defender\Module\Scan\Model\Settings::instance()->notification,
					'time'         => \WP_Defender\Module\Scan\Model\Settings::instance()->time,
					'day'          => \WP_Defender\Module\Scan\Model\Settings::instance()->day,
					'frequency'    => \WP_Defender\Module\Scan\Model\Settings::instance()->frequency
				),
				'audit_enabled'      => false,
				'audit_page_url'     => network_admin_url( 'admin.php?page=wdf-logging' ),
				'labels'             => $labels,
				'scan_page_url'      => network_admin_url( 'admin.php?page=wdf-scan' ),
				'hardener_page_url'  => network_admin_url( 'admin.php?page=wdf-hardener' ),
				'new_scan_url'       => network_admin_url( 'admin.php?page=wdf-scan&wdf-action=new_scan' ),
				'schedule_scans_url' => network_admin_url( 'admin.php?page=wdf-schedule-scan' ),
				'settings_page_url'  => network_admin_url( 'admin.php?page=wdf-settings' )
			) ),
		);

		$end_point = "https://premium.wpmudev.org/api/defender/v1/scan-results";
		$this->devCall( $end_point, $data, array(
			'method' => 'POST'
		) );
	}

	/**
	 * @param null $result
	 *
	 * @return bool
	 */
	public function checkRequirement( &$result = null ) {
		$meet = true;

		if ( version_compare( $this->getPHPVersion(), '5.3', '>=' ) == false ) {
			$meet          = false;
			$result['php'] = array(
				'status'  => $this->getPHPVersion(),
				'message' => __( "Please upgrade to 5.3 or later", wp_defender()->domain )
			);
		}

		return $meet;
	}


	/**
	 * @param $freq
	 *
	 * @return string
	 */
	public function frequencyToText( $freq ) {
		$text = '';
		switch ( $freq ) {
			case 1:
				$text = __( "daily", wp_defender()->domain );
				break;
			case 7:
				$text = __( "weekly", wp_defender()->domain );
				break;
			case 30:
				$text = __( "monthly", wp_defender()->domain );
				break;
		}

		return $text;
	}


	/**
	 * List server types
	 *
	 * @return Array
	 */
	public function serverTypes() {
		return apply_filters( 'defender_server_types', array(
			'apache'    => 'Apache',
			'litespeed' => 'LiteSpeed',
			'nginx'     => 'NGINX',
			'iis'       => 'IIS',
			'iis-7'     => 'IIS 7'
		) );
	}
}